### DISCUSSION TITLE
What does `-host-ip` and `-host-name -` mean for `server-tls`?

### DISCUSSION THREAD

**Discussion by McSinyx** [pymumu/smartdns#1591] [Q&A] [answered]
The [rich text documentation for the configuration format](https://pymumu.github.io/smartdns/en/configuration) lists `-host-ip` as an option for `server-tls`.  Does it override the server IP argument earlier on?  This page is also inconsistent with the sample `smartdns.conf`, which lists `-host-ip` for `server`.  Which one is correct?

As for the SNI hostname, is specifying `-host-name -` for `server-tls` or `server-https` any different from omitting the option?

**Comment by pymumu** [comment, marked as answer]
1. 
The `-host-ip` parameter is used to explicitly set the IP address to the upstream server to avoid bootstrap query, similar to /etc/hosts.
For example, `server-tls dns.google -host-ip=1.2.3.4` actually same as `server-tls 1.2.3.4 -hostname dns.google`

2. `-host-name` corresponds to the hostname of tls, which is SNI.
set `-host-name` to `-` actually lets smartdns query without setting SNI, the parameter takes effect on DOT and DOH servers.
For daily use, most of these parameters will be set according to the domain name. Just configure the upstream server using the URL format, such as
```
server-tls dns.google
server https://dns.google
```

The following three configurations are actually the same.

`server tls://dns.google -host-ip 8.8.8.8` 
`server-tls dns.google -host-ip 8.8.8.8` 
`server-tls 8.8.8.8 -host-name dns.google -tls-host-verify dns.google`

These parameters are also the same for `server-https`

`server https://dns.google -host-ip 8.8.8.8` 
`server-https https://dns.google -host-ip 8.8.8.8` 
`server-https 8.8.8.8 -host-name dns.google -tls-host-verify dns.google -host-name dns.google`

**Reply by McSinyx** [reply]
I see, so `server-tls` (and only the TLS one because I suppose it doesn't make sense for e.g. TCP) also take in a hostname as an argument, alternative to `ip[:port]`?

**Comment by PikuZheng** [comment]
This is related to TLS verification. Some server certificates do not have an IP address, which causes the verification to fail when connecting using the IP address. At this time, you have to specify the hostname to pass the authentication. 
hostip so that you don't have to resolve the hostname first when connecting.
